torch/csrc/tensor/fake_tensor.cpp

```diff
@@ -0,0 +1,273 @@
+
+using namespace pybind11::literals;
+
+namespace std {
+template <>
+struct hash<c10::SymInt> {
+  size_t operator()(const c10::SymInt& si) const noexcept {
+    if (auto mi = si.maybe_as_int()) {
+      return std::hash<int64_t>{}(*mi);
+    } else {
+      // Use the SymNode address as the hash.
+      auto b = si.toSymNodeImplUnowned();
+      return std::hash<void*>{}(b);
+    }
+  }
+};
+
+} // namespace std
+
+namespace {
+
+using SerializeBuffer = std::vector<uint8_t>;
+
+void serialize(SerializeBuffer& result, uint64_t value) {
+  result.push_back((uint8_t)value);
+  result.push_back((uint8_t)(value >> 8));
+  result.push_back((uint8_t)(value >> 16));
+  result.push_back((uint8_t)(value >> 24));
+  result.push_back((uint8_t)(value >> 32));
+  result.push_back((uint8_t)(value >> 40));
+  result.push_back((uint8_t)(value >> 48));
+  result.push_back((uint8_t)(value >> 56));
+}
+
+void serialize(SerializeBuffer& result, int8_t value) {
+  result.push_back((uint8_t)value);
+}
+
+void serialize(SerializeBuffer& result, int16_t value) {
+  result.push_back((uint8_t)value);
+  result.push_back((uint8_t)(value >> 8));
+}
+
+void serialize(SerializeBuffer& result, int32_t value) {
+  result.push_back((uint8_t)value);
+  result.push_back((uint8_t)(value >> 8));
+  result.push_back((uint8_t)(value >> 16));
+  result.push_back((uint8_t)(value >> 24));
+}
+
+void serialize(SerializeBuffer& result, int64_t value) {
+  serialize(result, (uint64_t)value);
+}
+
+void serializePointer(SerializeBuffer& result, void* value) {
+  serialize(result, (uint64_t)value);
+}
+
+// General enum serializer.
+template <
+    typename T,
+    typename = typename std::enable_if<std::is_enum<T>::value, T>::type>
+void serialize(SerializeBuffer& result, T value) {
+  serialize(result, std::underlying_type_t<T>(value));
+}
+
+void serialize(SerializeBuffer& result, const c10::SymInt& value) {
+  if (auto i = value.maybe_as_int()) {
+    result.push_back(0);
+    serialize(result, *i);
+  } else {
+    result.push_back(1);
+    serializePointer(result, value.toSymNodeImplUnowned());
+  }
+}
+
+void serialize(SerializeBuffer& result, const std::string& s) {
+  result.insert(result.end(), s.begin(), s.end());
+}
+
+void serialize(SerializeBuffer& result, c10::Device value) {
+  serialize(result, value.str());
+}
+
+template <typename T>
+void serialize(SerializeBuffer& result, const c10::ArrayRef<T>& value) {
+  serialize(result, value.size());
+  for (const auto& v : value) {
+    serialize(result, v);
+  }
+}
+
+struct Globals {
+  static const Globals& singleton() {
+    // Purposely leak this object!
+    static Globals* instance = new Globals();
+    return *instance;
+  }
+
+  Globals(const Globals&) = delete;
+  Globals& operator=(Globals&) = delete;
+
+ private:
+  Globals() = default;
+};
+
+const std::array<int, 4> DIM_ORDER_4{1, 3, 2, 0};
+const std::array<int, 5> DIM_ORDER_5{1, 4, 3, 2, 0};
+
+bool are_strides_like_channels_last(const at::Tensor& tensor) {
```

Comment:
If you're looking for a better review on this part, lemme know. I never looked deeply into channels last format. So I have no idea. Is this the sort of thing that deserves some unit tests? I don't know what precedent we have for testing c++ bindings like this.

Reply:
I guess the same for the functionality below that gets into the "internals" like is_sparse